Speed up /jobs for salt-api when run under cherrypy.

sthrasher opened the pull request:
### What does this PR do?

This PR increases performance of the /jobs endpoint of rest_cherrypy by eliminating a case where 2 salt runners are executed where only one is necessary.
### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Currently, if a jid is supplied, two runners are run (both
jobs.lookup_jid and jobs.list_job). This is unnecessary because
the return from jobs.list_job contains all of the information
from jobs.lookup_jid as well.

This change runs only jobs.list_job in this case and builds an
appropriate return from the output, netting an ~ 40% decrease
in response time for this endpoint when a jid is supplied.

sthrasher committed:
Speed up /jobs for salt-api when run under cherrypy.

Currently, if a jid is supplied, two runners are run (both
jobs.lookup_jid and jobs.list_job). This is unnecessary because
the return from jobs.list_job contains all of the information
from jobs.lookup_jid as well.

This change runs only jobs.list_job in this case and builds an
appropriate return from the output, netting an ~ 40% decrease
in response time for this endpoint when a jid is supplied.

whiteinge commented on salt/netapi/rest_cherrypy/app.py:
```diff
@@ -1210,29 +1210,21 @@ def GET(self, jid=None, timeout=''):
-            job_ret, job_info = job_ret_info
-            ret['info'] = [job_info]
+            ret['info'] = job_ret_info[0]
+            ret['return'] = [dict((k, job_ret_info[0]['Result'][k]['return']) for k in job_ret_info[0]['Result'])]
```
I'm not sure I follow the intent of this list comprehension. Would the following suffice?

```
job_info = job_ret_info[0]
job_ret = job_info.get('Result', {})
[...]
ret['return'] = [job_ret]
```

sthrasher replied:
The intent is to preserve the original format for clients of the API that
might depend on it. The comprehension converts the contents of the Results
key from a lookup_job query to be consistent with the results from
lookup_jid; so from:

Result:
    ----------
    host1:
        ----------
        return:
            True
    host2:
        ----------
        return:
            True

to:

host1:
    True
host2:
    True





On Mon, Aug 22, 2016 at 11:05 AM, Seth House <notifications@github.com>
wrote:

> In salt/netapi/rest_cherrypy/app.py
> <https://github.com/saltstack/salt/pull/35665#discussion_r75727517>:
>
> >          cherrypy.request.lowstate = lowstate
> >          job_ret_info = list(self.exec_lowstate(
> >              token=cherrypy.session.get('token')))
> >
> >          ret = {}
> >          if jid:
> > -            job_ret, job_info = job_ret_info
> > -            ret['info'] = [job_info]
> > +            ret['info'] = job_ret_info[0]
> > +            ret['return'] = [dict((k, job_ret_info[0]['Result'][k]['return']) for k in job_ret_info[0]['Result'])]
>
> I'm not sure I follow the intent of this list comprehension. Would the
> following suffice?
>
> job_info = job_ret_info[0]
> job_ret = job_info.get('Result', {})
> [...]
> ret['return'] = [job_ret]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/35665/files/25b43befaa5183b3e6da07066732cdc8e10689dc#r75727517>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHKLyTdGe6bALils2XTakSnUsqG2x3v0ks5qieT-gaJpZM4JqHAh>
> .
>

cachedout replied:
Ping @whiteinge for a follow-up.

whiteinge replied:
@sthrasher I apologize for the slow reply.

Thanks for the explanation. You're right, I overlooked the extra `return` key in that runner output. There's one remaining change needed to keep the existing return format: The top-level `return` is a single-item list containing the job return dictionary. This is true of the top-level `info` key in the job detail return.

E.g.,

Job list:
```json
{
  "return": [
  {
    "<jid>": {...job info...},
    "<jid>": {...job info...},
  }
  ]
}
```

Job detail:
```json
{
  "info": [
    {...job info...},
  ],
  "return": [
    {...job returns...},
  ]
}
```

whiteinge commented:
@sthrasher I like where you're headed with this PR. A 40% decrease in response time is pretty killer. It would nice if runners were faster, but that's a separate concern. 😛 

One minor compatibility note above.

One other thing I would suggest is to code defensively when reformatting that runner output. Right now salt-api is blindly returning whatever it gets from Salt but this change would require salt-api to depend on a certain return "shape" and those jobs runners do not currently have any test coverage. (We sorely need to get those runners tested.) The same concern but from elsewhere is that return shape can sometimes be inconsistent when using the `external_job_cache` or `master_job_cache` settings with a returner module that wasn't written to match the normal return shape.

If the return shape is not what we're expecting salt-api can **not** traceback.

sthrasher committed:
Fix compat issues with /jobs return values.

* ensure that both the return and info keys are lists.
* When the Result key doesn't match the expected shape,
  return an empty list of an empty dict (this matches existing
  behavior when looking for a non-existant JID).

sthrasher committed:
Switch from comprehension to logic used in jobs runner. This makes it
easier to deal with potential unicode in returns.

sthrasher commented:
I believe I've fixed both issues, the format of the return from salt-api
(managed to miss the info section being a list), as well as the return
shape issue.

On Tue, Aug 30, 2016 at 12:22 AM, Seth House <notifications@github.com>
wrote:

> @sthrasher <https://github.com/sthrasher> I like where you're headed with
> this PR. A 40% decrease in response time is pretty killer. It would nice if
> runners were faster, but that's a separate concern. 😛
>
> One minor compatibility note above.
>
> One other thing I would suggest is to code defensively when reformatting
> that runner output. Right now salt-api is blindly returning whatever it
> gets from Salt but this change would require salt-api to depend on a
> certain return "shape" and those jobs runners do not currently have any
> test coverage. (We sorely need to get those runners tested.) The same
> concern but from elsewhere is that return shape can sometimes be
> inconsistent when using the external_job_cache or master_job_cache
> settings with a returner module that wasn't written to match the normal
> return shape.
>
> If the return shape is not what we're expecting salt-api can *not*
> traceback.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/35665#issuecomment-243354803>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHKLyUap5nMKFVG7Y9nmPPVn0NTonJncks5qk9pJgaJpZM4JqHAh>
> .
>

whiteinge commented:
Solid work! Thanks so much for identifying the redundant work and then coding up a better solution. This looks like your first Salt contribution too. Welcome! 😄 

@cachedout I'll defer to you for the merge since there are (unrelated) test failures and I'm a little out-of-the-loop on that workflow now.
